bids/layout/layout.py

```diff
@@ -113,8 +113,23 @@ def __init__(self, root, validate=True, index_associated=True,
 
         # Validate arguments
         if not isinstance(root, six.string_types):
-            raise ValueError("root argument must be a string specifying the"
-                             " directory containing the BIDS dataset.")
+            # attempt to handle pathlib paths before giving up
+            try:
+                import pathlib
+                assert isinstance(root, pathlib.Path)
```

Comment:
Don't you want to convert it to string since pybids below would expect it?

Reply:
I'd rather the user do this, if they insist on using pathlib in python <3.4... Kind of an edge case, no? Don't really want to change type of inputs if possible.

Reply:
I might be missing something, but how would the user do this? If they're passing in a string, the try/except block will never get triggered. If they pass a `Path`, they risk failure later when pybids expects a string. Doesn't this ensure failure on Py<3.6 when using pathlib?

Reply:
In a word, yes. In 3.6 the os modules started supporting pathlib inputs, so it's only practical to support pathlib from this version onward (unless we want to switch over to using pathlib in place of os internally, but then that's work, and we'd need to add it as a dependency).

So that's why the logic gets a little hairy. It would be good to do something more elegant, if anyone has any ideas.

Reply:
To be clear, it doesn't break things for anyone. There's no change for people who still use str type file paths. And for people who use pathlib.Path, this will now work if you're on 3.6+. And fail in a somewhat informative way if you're on older python versions. 

But I take your point that the benefit is limited right now. It's a first step to supporting what I guess will eventually be the standard way to work with file paths in python. It's turned out to be a lot more complicated than I thought so certainly no objections here if you want to take a pass for now. I don't have time to do a more fundamental refactor to use pathlib internally (assuming that's even what you want right now).

Reply:
Yeah, I totally see the long-term benefits if we were to do a bigger overhaul, but I don't think we can restrict support to >= 3.6 any time soon. I'm not entirely confident that allowing `Path` inputs won't break things in unexpected ways. As I noted below, this PR currently isn't testing anything other than `BIDSLayout` initialization, and one consequence of allowing a `Path` as `root` is that any downstream code that accesses `.root` has to be able to handle that, which is not at all guaranteed—and which we can't easily verify in any case. While it's probably not going to be a common or big issue, it seems unnecessary. If the only benefit to adding pathlib support right now is to make it so that someone using `Path` objects can pass them to `BIDSLayout` without having to think about it, then a reasonable compromise solution I'd be fine with is to follow @yarikoptic's suggestion and explicitly cast `Path` objects to strings. Would that work? That would also lay the groundwork for more comprehensive changes down the road.

Reply:
I think casing `Path` to string would work nicely, because if you cast a string to a string no harm is done. I using `pathlib` internally would be the ideal solution, but especially in light of `grabbit` handling most of the string paths, that would probably be a pretty big refactor. And I'm saying this is big fan of `pathlib`.

As an side, did we or did we not drop `2.x` support?

Reply:
we did not
;)

Reply:
Ok, we now cast to str, which means I could get rid of all these tricky version-dependent conditionals.